File: primes.rb
require 'time'

def get_primes7(n)
  return [] if n  < 2
  return [2] if n == 2

  # do only odd numbers starting at 3
  s = 3.upto(n + 1).select(&:odd?)

  mroot = n ** 0.5
  half = s.length
  i = 0
  m = 3
  until m > mroot do
    if s[i]
      j = (m * m - 3) / 2
      s[j] = nil
      until j >= half do
        s[j] = nil
        j += m
      end
    end
    i += 1
    m = 2 * i + 3
  end
  [2] + s.compact
end

startTime = Time.now.to_i
periodTime = ENV['RUN_TIME'].to_i

while (Time.now.to_i - startTime) < periodTime do
  res = get_primes7(10000000)
  puts "Found #{res.length} prime numbers."
end

File: primes.swift
import Foundation

func get_primes7(_ n: Int) -> [Int] {
  if n < 2 {
    return []
  } else if n == 2 {
    return [2]
  }

  // do only odd numbers starting at 3
  var s = Array(stride(from: 3, to: n, by: 2))

  let mroot: Int = Int(sqrt(Double(n)))
  let half = s.count
  var i = 0
  var m = 3
  while m <= mroot {
    if s[i] != 0 {
      var j: Int = (m*m - 3) / 2
      s[j] = 0
      while j < half {
        s[j] = 0
        j += m
      }
    }
    i += 1
    m = 2*i + 3
  }
  return [2] + s.filter { $0 != 0 }
}

if let period_time_var = ProcessInfo.processInfo.environment["RUN_TIME"] {
  let start_time = Date()
  let period_time = Int(period_time_var)!

  while Int(Date().timeIntervalSince(start_time)) < period_time {
    let res = get_primes7(10_000_000)
    print("Found \(res.count) prime numbers.")
  }
} else {
  print("RUN_TIME not found.")
}

